### PR TITLE
Promote Preview

### PR DESCRIPTION
**Pull request by jasoth**
Using this pull request to review the diffs and comment on them.

### CODE REVIEW

File: src/internal/Get-MsGraphResults.ps1 [hunk 1]
```diff
@@ -292,20 +338,46 @@ function Get-MsGraphResults {
...
+                            ## error while invoking graph
+                            if ($Retries -eq $MaxRetries-1) {
+                                # catch error if it was the last try
+                                Catch-MsGraphError $_ 
+                            }
+                            ## check if throttling happened
+                            if ($_.Exception.PSobject.Properties.Name.Contains("Response") -and $_.Exception.Response.StatusCode.value__ -eq 429) {
+                                # Get the retry after header
+                                try {
+                                    $RetryAfter = [double]($_.Exception.Response.Headers.GetValues('Retry-After')[0])
```
Comment (jasoth) [inline src/internal/Get-MsGraphResults.ps1:357]: Has this been tested on both PS7 and Windows PS 5.1? I know the exception object is different so just wanting to double check.

File: src/internal/Get-MsGraphResults.ps1 [hunk 2]
```diff
@@ -433,7 +530,32 @@ function Get-MsGraphResults {
                             $uriQueryEndpointUniqueId.Path = ([IO.Path]::Combine($uriQueryEndpointUniqueId.Path, $id))
                         }
                         if ($DisableUniqueIdDeduplication -or $hashUri.Add($uriQueryEndpointUniqueId.Uri)) {
-                            if (!$DisableGetByIdsBatching -and $id -match '^[{]?[0-9a-fA-F]{8}-([0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12}[}]?$' -and $uriQueryEndpoint.Uri.Segments.Count -eq 3 -and $uriQueryEndpoint.Uri.Segments[2] -in ('directoryObjects', 'users', 'groups', 'devices', 'servicePrincipals', 'applications') -and ($QueryParametersFinal.Count -eq 0 -or ($QueryParametersFinal.Count -eq 1 -and $QueryParametersFinal.ContainsKey('$select')))) {
+                            if ($EnableInFilter -and $id -match '^[{]?[0-9a-fA-F]{8}-([0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12}[}]?$') {
```
Comment (jasoth) [inline src/internal/Get-MsGraphResults.ps1:533]: The long-term goal would be to enable the In Filter method by default right? Any objection to inverting this switch to follow the same pattern as the other batching switches? So something like "$DisableInFilterBatching". If it is still too early to make it the default, then perhaps we put a not in the parameter description that this parameter is experimental and may change?

File: src/Invoke-AADAssessmentDataCollection.ps1 [hunk 3]
```diff
@@ -239,70 +207,91 @@ function Invoke-AADAssessmentDataCollection {
...
-        $ReferencedIdCache.user | Get-MsGraphResults 'users?$select=id,userPrincipalName,userType,displayName,accountEnabled,onPremisesSyncEnabled,onPremisesImmutableId,mail,otherMails,proxyAddresses,assignedPlans' -TotalRequests $ReferencedIdCache.user.Count -DisableUniqueIdDeduplication -ApiVersion 'beta' `
+        # get user information
+        #$ReferencedIdCache.user | Get-MsGraphResults 'users/{0}?$select=id,userPrincipalName,userType,displayName,accountEnabled,onPremisesSyncEnabled,onPremisesImmutableId,mail,otherMails,proxyAddresses,assignedPlans,signInActivity' -TotalRequests $ReferencedIdCache.user.Count -DisableUniqueIdDeduplication -ApiVersion 'beta' `
+        $userQuery = 'users/{0}?$select=id,userPrincipalName,userType,displayName,accountEnabled,onPremisesSyncEnabled,onPremisesImmutableId,mail,otherMails,proxyAddresses'
+        if (!$NoAssignedPlans) {
+            $userQuery += ",assignedPlans"
```
Comment (jasoth) [inline src/Invoke-AADAssessmentDataCollection.ps1:284]: This method is fine for now. But we really need to do some more in-depth memory testing to see where the data collection and report generation processes can be improved for large tenants. I can help with that work after we complete this merge.

**Review by jasoth — COMMENTED** [review]
The changes to MS Graph look fine as far as I can tell without testing it. If it appears to be working, then it should be fine to merge. We can always go back and improve things further if required. I made a few other comments as well but nothing super important. I would be fine with merging everything as is right now if there is disagreement on my suggested changes. Again, we can always go back and change things more for the next version.

File: src/AzureADAssessment.psd1 [hunk 4]
```diff
@@ -116,10 +115,13 @@ NestedModules = @(
     '.\Get-AADAssessAppCredentialExpirationReport.ps1'
     '.\Get-AADAssessConsentGrantReport.ps1'
     '.\Get-AADAssessNotificationEmailsReport.ps1'
+    '.\Get-AADAssessRoleAssignmentReport.ps1'
+    '.\Get-AADAssessUserReport.ps1'
     '.\Invoke-AADAssessmentDataCollection.ps1'
     '.\New-AADAssessmentRecommendations.ps1'
     '.\Export-AADASsessmentRecommendations.ps1'
     '.\Import-AADAssessmentEvidence.ps1'
+    '.\Export-AADAssessmentReportData.ps1'
```
[the remark this answers is not in the document]
Comment (cblomart) [reply src/AzureADAssessment.psd1:124]: Granted this is mainly for exceptional cases where reports don't follow through correctly for ressource reasons (time included). 
Having the command separately allows to redo the reporting from where it failed hopefully giving an alternative to redoing the full collection of data.
Thinking about it, if kept as is we may include a message that collection is properly finished and that the next step are reports allowing users/assessors to know that if they encounter an error they can redo the reporting only.

File: src/internal/Get-MsGraphResults.ps1 (hunk 1 above)
Comment (cblomart) [reply src/internal/Get-MsGraphResults.ps1:357]: i tested on PS7 only will validate on PS5.1

File: src/internal/Get-MsGraphResults.ps1 (hunk 2 above)
Comment (cblomart) [reply src/internal/Get-MsGraphResults.ps1:533]: I would propose to keep the enable and include the experimental warning. As soon as we can set that as "not experimental" we can switch to a DisableInFilter behavior. Will follow up with Graph team.

File: src/internal/Get-MsGraphResults.ps1 (hunk 1 above)
Comment (cblomart) [reply src/internal/Get-MsGraphResults.ps1:357]: Brought some updates for issues seen on PS5.1 (mostly arround app credential & certs)
Didn't have specific throttling. in this code if the Exception doesn't have a "Response" it should fallback to the default timings.

**Review by cblomart — APPROVED** [review]

**Pull request merged**